library/java/net/openid/appauth/AuthorizationService.java

```diff
@@ -409,26 +424,162 @@ protected void onPostExecute(JSONObject json) {
+        private ClientAuthentication mClientAuthentication;
+
+        private AuthorizationException mException;
+
+        TokenValidationRequestTask(TokenResponse request, @NonNull ClientAuthentication clientAuthentication,
+                                   TokenValidationResponseCallback callback) {
+            response = request;
+            mCallback = callback;
+            mClientAuthentication = clientAuthentication;
+        }
+
+        @Override
+        protected JSONObject doInBackground(Void... voids) {
+            InputStream is = null;
+            try {
+                Uri uri = (response.request.configuration.discoveryDoc.getJwksUri());
+                HttpURLConnection conn =
+                        mClientConfiguration.getConnectionBuilder()
+                                .openConnection(uri);
+                conn.setRequestMethod("GET");
+//                conn.setDoOutput(true);
+
+                is = new BufferedInputStream(conn.getInputStream());
+                // readStream(in);
+                String response = Utils.readInputStream(is);
+                return new JSONObject(response);
+            } catch (IOException ex) {
+                Logger.debugWithStack(ex, "Failed to complete exchange request");
+                mException = AuthorizationException.fromTemplate(
+                        GeneralErrors.NETWORK_ERROR, ex);
+            } catch (JSONException ex) {
+                Logger.debugWithStack(ex, "Failed to complete exchange request");
+                mException = AuthorizationException.fromTemplate(
+                        GeneralErrors.JSON_DESERIALIZATION_ERROR, ex);
+            } finally {
+                Utils.closeQuietly(is);
+            }
+            return null;
+        }
+
+        @Override
+        protected void onPostExecute(JSONObject json) {
+            JSONObject requiredJson = null;
+
+            try {
+                String[] split = response.idToken.split("\\.");
+                Log.d("JWT_DECODED", "Header: " + JsonUtil.decodeBase64(split[0]));
+                Log.d("JWT_DECODED", "Body: " + JsonUtil.decodeBase64(split[1]));
+
+                String decodeToken_Header = JsonUtil.decodeBase64(split[0]);
+                String decodeToken_Body = JsonUtil.decodeBase64(split[1]);
+
+                Log.d("decodeToken ", decodeToken_Body);
+                JSONObject jsonObject_header = new JSONObject(decodeToken_Header);
+                JSONObject jsonObject_body = new JSONObject(decodeToken_Body);
+
+                JSONArray jsonArray = json.getJSONArray("keys");
+
+                for (int i = 0; i < jsonArray.length(); i++) {
+                    if (jsonArray.getJSONObject(i).optString("alg").contains(jsonObject_header.getString("alg"))) {
+                        requiredJson = jsonArray.getJSONObject(i);
+                        break;
+                    }
+                }
+
+                if (requiredJson != null) {
```

Comment:
Okay I will fix the nesting by inverting conditions and returning early as per your suggestions.